### ISSUE TITLE
Proposal: OpenTelemetry Tracing Per-Plugin 

### ISSUE THREAD

**Issue by Dentrax** [kubernetes-sigs/descheduler#951]
**Is your feature request related to a problem? Please describe.**

Tracing is one of the important observability topics about continuously watch overall health and performance of the system. Especially with the new descheduler framework, it would perfectly fit for implementing tracing per-plugin to observe such as `Deschedule()`, `Balance()`, etc. functions using spans.

**Describe the solution you'd like**

Implement [OpenTelemetry](https://opentelemetry.io/) `(go.opentelemetry.io)` client in Descheduler. [Here is an example](https://github.com/kyverno/kyverno/blob/9b41e2e0174c10b3548a5364e205911d6a0047c7/pkg/tracing/tracing.go) source file how Kyverno implemented OpenTracing in their codebase.

Introduce a new OpenTelemetry client in `framework.Handle` _interface_:
```
type Handle interface {
  ...
  TraceClient() *TracerProvider
}
```

Introduce a new boolean flag to `PodEvictor` struct to enable tracing:

```
type PodEvictor struct {
  ...
  tracingEnabled bool
}
```

Wrap all the following functions in a `DoInSpan` func if `tracingEnabled`, otherwise it should be as is:

Plugin initialization:
```
tracing.DoInSpan(context.Background(), "PLUGIN_NAME", "new", func(ctx context.Context) {
		pg, err := PLUGIN_NAME.New(args, handle)
})
```

Deschedule:
```
tracing.DoInSpan(context.Background(), "PLUGIN_NAME", "deschedule", func(ctx context.Context) {
		status = pg.(framework.DeschedulePlugin).Deschedule(ctx, nodes)
})
```

Balance:
```
tracing.DoInSpan(context.Background(), "PLUGIN_NAME", "balance", func(ctx context.Context) {
		status = pg.(framework.BalancePlugin).Balance(ctx, nodes)
})	
```

Tracing internal metrics of a custom plugin would be straightforward. Here is some function list that would consider worth to trace:

* `classifyNodes`
* `evictPodsFromSourceNodes`
* `getTargetNodes`
* observe end-start timing for entire `for i := 0; i < totalPods; i++ {` loops
* observe end-start timing for entire `for _, pod := range podsToEvict` loops
* observe end-start timing for entire `for _, node := range nodes` loops
* etc.

**Describe alternatives you've considered**
`-`

**What version of descheduler are you using?**
`-`

**Additional context**
`-`

[^1]: https://github.com/kubernetes-sigs/descheduler/pull/648#issuecomment-1089122875

cc @eminaktas

**Comment by harshanarayana** [comment]
I would love to take a look at this if no one else has already started on this.

**Comment by harshanarayana** [comment]
Does this also mean we have to enable the native apiserver tracing option via the `DeschedulerServer` ?

**Comment by Dentrax** [comment]
> native apiserver tracing option

I am not very familiar with that option. Do you have any article or reference for this?

**Comment by harshanarayana** [comment]
@Dentrax  https://github.com/kubernetes/apiserver/blob/3714462ec59d8c7263cb232141272489bb0566a8/pkg/server/options/recommended.go#L84. I am talking about the bits here. The library we use for for starting the descheduler has the tracing ability built into the core of this.

**Comment by Dentrax** [comment]
Wow, didn't notice, that's cool! Are there any code sample or proect that use `Traces` option?

**Comment by harshanarayana** [comment]
It needs a few thing to be done as a pre-requisite though. Setting `APIServerTracing` gate is one of them. Depending on which version you are on. 

```diff
diff --git a/cmd/descheduler/app/options/options.go b/cmd/descheduler/app/options/options.go
index 0d0a30ceb..d5ab81a54 100644
--- a/cmd/descheduler/app/options/options.go
+++ b/cmd/descheduler/app/options/options.go
@@ -42,6 +42,9 @@ type DeschedulerServer struct {
 	Client         clientset.Interface
 	EventClient    clientset.Interface
 	SecureServing  *apiserveroptions.SecureServingOptionsWithLoopback
+	Tracer         *apiserveroptions.TracingOptions
+	EgressSelector *apiserveroptions.EgressSelectorOptions
+	Feature        *apiserveroptions.FeatureOptions
 	DisableMetrics bool
 }
 
@@ -58,6 +61,9 @@ func NewDeschedulerServer() (*DeschedulerServer, error) {
 	return &DeschedulerServer{
 		DeschedulerConfiguration: *cfg,
 		SecureServing:            secureServing,
+		Tracer:                   apiserveroptions.NewTracingOptions(),
+		EgressSelector:           apiserveroptions.NewEgressSelectorOptions(),
+		Feature:                  apiserveroptions.NewFeatureOptions(),
 	}, nil
 }
 
@@ -93,4 +99,6 @@ func (rs *DeschedulerServer) AddFlags(fs *pflag.FlagSet) {
 	componentbaseoptions.BindLeaderElectionFlags(&rs.LeaderElection, fs)
 
 	rs.SecureServing.AddFlags(fs)
+	rs.Tracer.AddFlags(fs)
+	rs.Feature.AddFlags(fs)
 }
diff --git a/cmd/descheduler/app/server.go b/cmd/descheduler/app/server.go
index f060b3b94..1becf9514 100644
--- a/cmd/descheduler/app/server.go
+++ b/cmd/descheduler/app/server.go
@@ -32,6 +32,7 @@ import (
 
 	apiserver "k8s.io/apiserver/pkg/server"
 	"k8s.io/apiserver/pkg/server/mux"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	restclient "k8s.io/client-go/rest"
 	registry "k8s.io/component-base/logs/api/v1"
 	_ "k8s.io/component-base/logs/json/register"
@@ -56,11 +57,24 @@ func NewDeschedulerCommand(out io.Writer) *cobra.Command {
 
 			// LoopbackClientConfig is a config for a privileged loopback connection
 			var LoopbackClientConfig *restclient.Config
+			var apiserverConfig = &apiserver.Config{
+				LoopbackClientConfig: LoopbackClientConfig,
+			}
 			var SecureServing *apiserver.SecureServingInfo
 			if err := s.SecureServing.ApplyTo(&SecureServing, &LoopbackClientConfig); err != nil {
 				klog.ErrorS(err, "failed to apply secure server configuration")
 				return
 			}
+			apiserverConfig.LoopbackClientConfig = LoopbackClientConfig
+			if err := s.EgressSelector.ApplyTo(apiserverConfig); err != nil {
+				klog.ErrorS(err, "failed to apply egress selector configuration")
+				return
+			}
+			if err := s.Tracer.ApplyTo(apiserverConfig.EgressSelector, apiserverConfig); err != nil {
+				klog.ErrorS(err, "failed to apply trace configuration")
+				return
+			}
 			var factory registry.LogFormatFactory
 			if factory == nil {
 				klog.ClearLogger()
@@ -101,6 +115,7 @@ func NewDeschedulerCommand(out io.Writer) *cobra.Command {
 	cmd.SetOut(out)
 	flags := cmd.Flags()
 	s.AddFlags(flags)
+	utilfeature.DefaultMutableFeatureGate.AddFlag(flags)
 	return cmd
 }
 
```
Here is a hacked up bit of code that can be used to enable the API Server based tracing capabilities. 

But this is not entirely good enough for what is in question for the de-scheduler use-cases for example. But, if we enable tracing for the framework handler, should this be enabled too in conjunction ?

**Comment by harshanarayana** [comment]
Another thing I noticed is https://github.com/kubernetes/apiserver/blob/bdbf5341a8de58461fe16315330e7b67fbac520d/go.mod#L29, the version of `otel` here is really old. This might come in the way of adding the new otel sdk/traces libraries into this repo.

**Comment by harshanarayana** [comment]
@Dentrax  Thanks for the response

> Oh, this is tricky part then. I don't think people will mutate their API Server flags (e

They do not need to mutate the API Server Flags. Sorry, I meant to say Enable the Feature gate on the Descheduler service. Without which the tracer built into the kuberentes/apiserver package won't work. 

>  You don't have to mutate & restart all API servers, also you can manage the latest otel dependency.

Agreed. I definitely like the one in Kyverno better. More standalone in nature without external dependencies. 

> Unfortunately. [Latest](https://github.com/open-telemetry/opentelemetry-go/releases) release is v1.10.0. But Kubernetes uses v0.20, which was released almost 15+ months ago.

I tried to enable using the new version of the OTEL packages with the current code base of the scheduler But it doesn't like it because of the old dependency in the `kubernetes/apiserver` package. Any idea why do we even use that for the descheduler ?

**Comment by harshanarayana** [comment]
I see the `k8s.io/apiserver` dependency exists mainly for 

```bash
cmd/descheduler/app/server.go
26:	"k8s.io/apiserver/pkg/server/healthz"
33:	apiserver "k8s.io/apiserver/pkg/server"
34:	"k8s.io/apiserver/pkg/server/mux"
35:	utilfeature "k8s.io/apiserver/pkg/util/feature"
```
If we can get rid of that dependency, we can freely move to the right OTEL version of our liking.

**Comment by Dentrax** [comment]
I think the primary role of the `apiserver` is the serve Descheduler's [SecureServe](https://github.com/kubernetes-sigs/descheduler/blob/ea8e648cfb833ca27ce5636794a9cbf2233a1c9a/cmd/descheduler/app/options/options.go#L55) stuff. So removing could not be possible, I think.

> But it doesn't like it because of the old dependency in the kubernetes/apiserver package.

What about using `replace` directive in `go.mod`?

**Comment by harshanarayana** [comment]
> What about using replace directive in go.mod?

I did a quick try for that already. But looks like the structs around the package has changed considerably. 

```bash
go: finding module for package go.opentelemetry.io/otel/internal/metric
sigs.k8s.io/descheduler/cmd/descheduler/app/options imports
	k8s.io/apiserver/pkg/server/options imports
	go.opentelemetry.io/otel/semconv: module go.opentelemetry.io/otel@latest found (v0.0.0-00010101000000-000000000000, replaced by go.opentelemetry.io/otel@v1.7.0), but does not contain package go.opentelemetry.io/otel/semconv
sigs.k8s.io/descheduler/cmd/descheduler/app/options imports
	k8s.io/apiserver/pkg/server/options imports
	go.opentelemetry.io/otel/exporters/otlp/otlpgrpc imports
	go.opentelemetry.io/otel/exporters/otlp/internal/transform imports
	go.opentelemetry.io/otel/metric/number: module go.opentelemetry.io/otel/metric@latest found (v0.0.0-00010101000000-000000000000, replaced by go.opentelemetry.io/otel/metric@v0.30.0), but does not contain package go.opentelemetry.io/otel/metric/number
sigs.k8s.io/descheduler/cmd/descheduler/app/options imports
	k8s.io/apiserver/pkg/server/options imports
	go.opentelemetry.io/otel/exporters/otlp/otlpgrpc imports
	go.opentelemetry.io/otel/exporters/otlp/internal/transform tested by
	go.opentelemetry.io/otel/exporters/otlp/internal/transform.test imports
	go.opentelemetry.io/otel/sdk/metric/aggregator/exact: module go.opentelemetry.io/otel/sdk/metric@latest found (v0.0.0-00010101000000-000000000000, replaced by go.opentelemetry.io/otel/sdk/metric@v0.30.0), but does not contain package go.opentelemetry.io/otel/sdk/metric/aggregator/exact
sigs.k8s.io/descheduler/cmd/descheduler/app/options imports
	k8s.io/apiserver/pkg/server/options imports
	go.opentelemetry.io/otel/exporters/otlp/otlpgrpc imports
	go.opentelemetry.io/otel/exporters/otlp/internal/transform tested by
	go.opentelemetry.io/otel/exporters/otlp/internal/transform.test imports
	go.opentelemetry.io/otel/sdk/metric/aggregator/minmaxsumcount: module go.opentelemetry.io/otel/sdk/metric@latest found (v0.0.0-00010101000000-000000000000, replaced by go.opentelemetry.io/otel/sdk/metric@v0.30.0), but does not contain package go.opentelemetry.io/otel/sdk/metric/aggregator/minmaxsumcount
sigs.k8s.io/descheduler/cmd/descheduler/app imports
	k8s.io/apiserver/pkg/server imports
	k8s.io/apiserver/pkg/endpoints/filters imports
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp tested by
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.test imports
	go.opentelemetry.io/otel/oteltest imports
	go.opentelemetry.io/otel/internal/metric: module go.opentelemetry.io/otel@latest found (v0.0.0-00010101000000-000000000000, replaced by go.opentelemetry.io/otel@v1.7.0), but does not contain package go.opentelemetry.io/otel/internal/metric
sigs.k8s.io/descheduler/cmd/descheduler/app imports
	k8s.io/apiserver/pkg/server imports
	k8s.io/apiserver/pkg/endpoints/filters imports
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp tested by
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.test imports
	go.opentelemetry.io/otel/oteltest imports
	go.opentelemetry.io/otel/metric/registry: module go.opentelemetry.io/otel/metric@latest found (v0.0.0-00010101000000-000000000000, replaced by go.opentelemetry.io/otel/metric@v0.30.0), but does not contain package go.opentelemetry.io/otel/metric/registry
```

**Comment by damemi** [comment]
Hey @Dentrax @harshanarayana, thanks for proposing this! I'm actually on the OpenTelemetry team at google now, so I'm happy to help with any reviews or questions.

The SDK in the apiserver is outdated, and we definitely should be trying to use the GA stable version instead of that version. We were working on updating that version (which was actually blocked on etcd also being out of date). If we can't move that dependency, we might be blocked too.

There is more info on the API server's tracing [in the KEP for it](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/647-apiserver-tracing) (and a [blog post](https://kubernetes.io/blog/2021/09/03/api-server-tracing/)), but the idea is that any API calls we make can pass a context that lets users trace requests from descheduler->apiserver->other components. I think etcd and crio are the two main components that also have tracing implemented right now.

It would be really great if we can break that apiserver dependency and at least get tracing instrumented in our own codebase in the mean time. So that might be a good place to start

**Comment by harshanarayana** [comment]
@damemi  This is great info.  I went through the post and the KEP a while ago, and was incredibly useful for a personal work on the tracing bit. 

> It would be really great if we can break that apiserver dependency and at least get tracing instrumented in our own codebase in the mean time.

I was thinking the same. But, Like @Dentrax  mentioned, the `k8s.io/apiserver` package is present mainly to cater to the TLS requirements (from the code, that seems to be all it is meant for). But the code required to manage the cert is well built into the code of the `k8s.io/apiserver` under the `DynamicServingCertificateController`. Any idea what can be done to remove this dependency cleanly ?

**Comment by damemi** [comment]
> Also we need to pass tracing client into frameworks to trace each one individually.

There's the [global package](https://pkg.go.dev/go.opentelemetry.io/otel/global) that might work for us, it lets you register Providers for traces/metrics that you can use just by importing the package, so you don't need to pass them around as parameters.

**Comment by harshanarayana** [comment]
> the API server's otel dependencies were just updated to the latest versions in [kubernetes/kubernetes#112545](https://github.com/kubernetes/kubernetes/pull/112545) 🎉 so we should be unblocked now right?

This is wonderful

```bash
	go.opentelemetry.io/otel v1.10.0
	go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc v1.10.0
	go.opentelemetry.io/otel/sdk v1.10.0
	go.opentelemetry.io/otel/trace v1.10.0
```

All that we wanted. I will have a draft PR working on this if we are okay with me taking a stab at this. @Dentrax

**Comment by harshanarayana** [comment]
>  We can work collaboratively on this feature if you want! :)

@Dentrax Would be more than happy to work with you :)

**Comment by Dentrax** [comment]
@Dentrax Would be more than happy to work with you :)

🤗 I DM'ed you on Slack!

**Comment by harshanarayana** [comment]
@damemi @Dentrax  I am posting some of the questions and clarification here so that others can chime in as well for a better discussion. 

### Some Questions

1. What modes of exporter do we support ? `otel` has a few different one it already has support for in the form of `jaeger` `zipkin` etc along with the `otlmtracegrcp`. Would it be good to expose a way to let the end user pick the tracers via an argument and config corresponding to it ?
2. If the answer to q1 is yes, then would it be better to externalize the `otel` config into a file that can be provided as an input instead of a bunch of CLI arguments ? They are easy to read but not always user friendly to test and remember. 
3. What about metrics? Right now, I don't see major set of metrics being exposed. But do we keep some provision for this?
4. How do we extend the trace behavior into the plugins? Do we let each plugin use the helper function provided by the `pkg/tracing` ? 
    1.  I do like the idea of `DoInSpan` that `kyverno` has but I think the function being passed as an argument to that takes a good bit of control away from doing things in the code. 
5. If the answer to q4 is No, then how do we make the plugin code cleaner and keep it away from worrying about having to do too much around tracing ? A wrapper plugin that calls each plugin handler if tracer is enabled by setting the right span and terminating them?
6. Instead of worrying about if the tracer is enabled or not, can we simply setup a noop trace provider? Will keep the code cleaner as well I think.

#### Possible Approach for 4.1
For the Item 4.1 I think we can do a bit of a tweak and move away from `DoInSpan` but get all the advantages/behavior of it. 

```golang
func StartSpan(ctx context.Context, tracerName string, traceOpts []trace.TracerOption, operationName string, attributes []attribute.KeyValue) (context.Context, trace.Span, func()) {
	spanCtx, span := otel.Tracer(tracerName, traceOpts...).Start(ctx, operationName, trace.WithAttributes(attributes...))
	return spanCtx, span, func() {
		span.End()
	}
}
```

One would call the `tracing.StartSpan(.....)` and it returns a defferable anonymous function that can be used to terminate the span. This way, almost no code has to be changed other than adding a few lines in the beginning of each plugin functions that we need tracing. 

#### Possible Approach for 5
```golang
type TraceEnabledDeSchedulePlugin struct {
	plugin framework.DeschedulePlugin
}

func (t *TraceEnabledDeSchedulePlugin) Name() string {
	return t.plugin.Name()
}

func (t *TraceEnabledDeSchedulePlugin) Deschedule(ctx context.Context, nodes []*v1.Node) (status *framework.Status) {
	spanCtx, span, closer := StartSpan(ctx, t.plugin.Name(), []trace.TracerOption{}, "deschedule", []attribute.KeyValue{})
	defer closer()
	defer func() {
		if status.Err != nil {
			span.RecordError(status.Err)
		}
	}()
	return t.plugin.Deschedule(spanCtx, nodes)
}
```
This will just be a wrapper with no login in it other than extending the tracing capabilities.

**Comment by damemi** [comment]
> What modes of exporter do we support ?

We should only need to focus on OTLP support. Jaeger already supports OTLP and Otel has the [collector](https://opentelemetry.io/docs/collector/), which is the idomatic way to convert trace data to backends that have their own formats

> If the answer to q1 is yes, then would it be better to externalize the otel config into a file

if there is otel specific config we want to offer, then it should just be part of the descheduler's componentconfig API. right now the descheduler's config is mostly plugin based, but we want to make it hold more component settings too (https://github.com/kubernetes-sigs/descheduler/pull/857#discussion_r897804632)

> What about metrics? Right now, I don't see major set of metrics being exposed. But do we keep some provision for this?

we don't have many metrics right now, but it would be nice to add some more such as latency, failure rates, etc. But the otel go metrics sdk is just barely in alpha, so we might want to wait on adopting otel for that. On the other hand, if we want to try out the alpha sdk to help that along then I'm all for it. But that should be a separate issue

> How do we extend the trace behavior into the plugins?

I think the designs are a bit more than is necessary for this. It should be fine for plugins to handle tracing themselves because of how lightweight the tracing sdk is. It should be simple enough to drop-in the couple lines of code for starting and ending a trace in the plugin's logic. 

The descheduler framework can register the global tracer provider for them to use, but we don't need to enforce this on plugin developers. We already pass a `context` to each plugin, which should be enough info for them to tie in their own spans.

**Comment by Dentrax** [comment]
I made some changes to PR https://github.com/kubernetes-sigs/descheduler/pull/958/files:

My commit is: 4d1bdfc040cb30f430fd4669fa843ebb7d2bf4bc

**Working Demo**

```bash
$ make
$ ./_output/bin/descheduler --policy-config-file ./examples/policy.yaml --leader-elect=true --kubeconfig ~/kubeconfig --descheduling-interval 10s -v 5 --otel-endpoint 10.11.12.13:55680
# wait a few mins
```

<img width="1714" alt="Screen Shot 2022-09-30 at 15 18 42" src="https://user-images.githubusercontent.com/16493751/193267949-fbfd40a8-64f2-45e9-afc8-c36fef832090.png">

@damemi I had to introduce a new function called `RunPlugin`. I iterated Harsha's second possible approach a bit further and create a common function. Please take a look!

We can start write tests once we get final design. 🤞 

PTAL @harshanarayana @eminaktas

**Comment by harshanarayana** [comment]
```bash
❯ ./hack/update-generated-conversions.sh
E1003 23:08:34.766133   48655 conversion.go:756] Warning: could not find nor generate a final Conversion function for sigs.k8s.io/descheduler/pkg/apis/componentconfig.DeschedulerConfiguration -> sigs.k8s.io/descheduler/pkg/apis/componentconfig/v1alpha1.DeschedulerConfiguration
E1003 23:08:34.766317   48655 conversion.go:757]   the following fields need manual conversion:
E1003 23:08:34.766347   48655 conversion.go:759]       - Tracing

```

@damemi  @Dentrax  I was trying to run the `make verify` and keep running into this. This seem to be a side effect of adding the new `Tracing` attribute to the `DeschedulerConfiguration`. I am unable to figure out exactly what needs to be done to resolve this. Any help on this would be really great.

**Comment by damemi** [comment]
@harshanarayana it's probably because the PR adds it to [`pkg/apis/componentconfig/types.go`](https://github.com/kubernetes-sigs/descheduler/pull/958/files#diff-8cb3a832fd89345b8caf52788b477463aaf7cd78c7a947d55eef89fdc1d5e12f) but not [`pkg/apis/componentconfig/v1alpha1/types.go`](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/apis/componentconfig/v1alpha1/types.go).

It may or may not ultimately end up in the componentconfig as we work out the v1alpha2 Policy/config api

**Comment by Dentrax** [comment]
Run from Harsha's latest commit: d37ab93f974c067664c3815db51c944e7e2a613f 🚀 

![Screen Shot 2022-10-04 at 14 03 13](https://user-images.githubusercontent.com/16493751/193803105-ff48cd6f-d709-4f31-a5c1-b204e6cd5e11.png)

**Comment by damemi** [comment]
/lifecycle frozen
this is being actively worked on

**Comment by harshanarayana** [comment]
Changes merged in #1189 

/close
